### DISCUSSION TITLE
using risor for templating

### DISCUSSION THREAD

**Discussion by newbeelearn** [deepnoodle-ai/risor#68] [Ideas]
I am always on lookout for a good templating system in golang. previously i evaluated scriggo which tick marked everything except there was no easy way to access json. Golang templates work fine for smalll use cases but severly lack in comparison to other languages and fail short by changing the basic golang syntax for e.g. using range.

My use case is very simple i want to take json file and template as input, execute template and write it in output file. Any example covering this use case would be great.

**Comment by myzie** [comment]
Hi @newbeelearn. I've thought a bit about how to make templating easy with Risor, but hadn't made a decision on this yet.

For simple string templates, Risor does support string interpolation like [Python f-strings](https://realpython.com/python-f-strings/#simple-syntax).

You can see that in action with a Risor CLI command like this:

```bash
$ risor -c "name := 'alice'; print('hi there {name}')"
hi there alice
```

For more complex templates, especially ones requiring range/loops then I would anticipate either adding support for Go's standard library templates or one of the popular open source templating libraries. But it would need to be one that doesn't require the templates to be compiled into Go (so ones like https://github.com/valyala/quicktemplate won't work here).

It seems like either https://github.com/flosch/pongo2 or the Go standard library ones, from my quick look.

Can you share an example template so I can understand your use case better?

**Reply by luisdavim** [reply]
I'd prefer to stick to the Go standard library on this one.

**Comment by luisdavim** [comment]
@newbeelearn can you elaborate a bit more on your use cases for templating?

@myzie I'm thinking, since risor is aimed for devops scripting, having something that helps generating config files, k8s manifests and api payloads, mostly yaml and json... would be useful.

I could open a PR to add a templating module, using Go's standard lib templating and some additional functions to make it similar to the functionality that helm provides. I think helm is pretty good at manipulating yaml and json both as inputa and output but can also be used to render other file formats. What do you think?

**Reply by myzie** [reply]
100% agree. I'm in favor of starting with the standard lib's templating as you said. The YAML I happen to be generating for work right now is Docker Swarm files (of all things!). I've only used Helm a bit, so sounds like you're more familiar with it. Does Helm internally use vanilla Go templates?

**Reply by newbeelearn** [reply]
so i generate lot of my CRUD code using templates. This includes front end code, back-end handlers and sql queries etc. You can think of it as generic https://postgrest.org/en since it can work with any database not limited to postgres.

Indeed helm is the right example of what i had in mind.

**Reply by luisdavim** [reply]
Yes, helm uses go vanilla templating with some helper functions, I can get pretty close to what it provides without much effort, I'll start a PR and we can take it from there.

**Reply by luisdavim** [reply]
here's the PR, please let me know what you think: https://github.com/risor-io/risor/pull/152

I'm using a pipe in the example script but you can also pass the values for the template as the first argument:

```
#!/usr/bin/env risor --

data := fetch("http://ipinfo.io").json()
res := template(data, "You are in {{ .Values.city }}, region {{ .Values.region }} in {{ .Values.timezone }}")

print(res)
```

**Reply by luisdavim** [reply]
@myzie I've kept the k8s helper function separate and under a flag, I think that from go 1.17 lazy loading is supported and that should keep the binary small if the flag is not set but I'm not 100% sure on that...

**Comment by luisdavim** [comment]
@newbeelearn like helm, this uses https://github.com/Masterminds/sprig so pretty much all the functions mentioned in the helm docs should be available, I've also added some that where missing like toYaml for and fromYaml and k8slookup should be similar to the lookup function from helm, let me know if you find something is missing.
